File: 48_positional_encoding_timeseries/python/examples/basic_encoding.py
#!/usr/bin/env python3
"""
Basic Positional Encoding Example

This example demonstrates the fundamental positional encoding types:
sinusoidal, learned, relative, and rotary (RoPE).
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import torch
import numpy as np
from positional_encoding import (
    SinusoidalPositionalEncoding,
    LearnedPositionalEncoding,
    RelativePositionalEncoding,
    RotaryPositionalEncoding,
)


def main():
    print("Basic Positional Encoding Example")
    print("=" * 50)

    d_model = 64
    max_len = 100
    batch_size = 2
    seq_len = 10

    # 1. Sinusoidal Encoding
    print("\n1. Sinusoidal Positional Encoding")
    print("-" * 40)

    sinusoidal = SinusoidalPositionalEncoding(d_model, max_len)
    x = torch.zeros(batch_size, seq_len, d_model)
    encoded = sinusoidal(x)

    print(f"Input shape: {x.shape}")
    print(f"Output shape: {encoded.shape}")
    print(f"Position 0, dims 0-3: {encoded[0, 0, :4].numpy().round(4)}")
    print(f"Position 1, dims 0-3: {encoded[0, 1, :4].numpy().round(4)}")
    print("Note: At position 0, sin(0)=0 and cos(0)=1 pattern")

    # 2. Learned Encoding
    print("\n2. Learned Positional Encoding")
    print("-" * 40)

    learned = LearnedPositionalEncoding(d_model, max_len)
    encoded = learned(x)

    print(f"Trainable parameters: {sum(p.numel() for p in learned.parameters())}")
    print(f"Output shape: {encoded.shape}")
    print(f"Embedding norm at pos 0: {torch.norm(learned.embedding.weight[0]).item():.4f}")
    print("Note: These would be trained during model training")

    # 3. Relative Encoding
    print("\n3. Relative Positional Encoding")
    print("-" * 40)

    n_heads = 4
    head_dim = d_model // n_heads
    relative = RelativePositionalEncoding(d_model, n_heads, max_relative_position=50)

    # Create Q, K, V tensors [batch, n_heads, seq_len, head_dim]
    q = torch.randn(batch_size, n_heads, seq_len, head_dim)
    k = torch.randn(batch_size, n_heads, seq_len, head_dim)
    v = torch.randn(batch_size, n_heads, seq_len, head_dim)

    # Apply relative attention
    output = relative(q, k, v)

    print(f"Q/K/V shape: [batch={batch_size}, heads={n_heads}, seq={seq_len}, head_dim={head_dim}]")
    print(f"Output shape: {output.shape}")
    print(f"Max relative position: 50")
    print("Note: Modifies attention scores based on relative positions")

    # 4. Rotary Encoding (RoPE)
    print("\n4. Rotary Positional Encoding (RoPE)")
    print("-" * 40)

    # RoPE expects [batch, n_heads, seq_len, head_dim]
    rope = RotaryPositionalEncoding(d_model, n_heads, max_len=max_len)
    q_rope = torch.randn(batch_size, n_heads, seq_len, head_dim)
    k_rope = torch.randn(batch_size, n_heads, seq_len, head_dim)

    q_rot, k_rot = rope(q_rope, k_rope)

    print(f"Query/Key shape: [batch={batch_size}, heads={n_heads}, seq={seq_len}, head_dim={head_dim}]")
    print(f"Rotated Q/K shape: {q_rot.shape}")

    # Verify norm preservation
    q_norm_before = torch.norm(q_rope, dim=-1).mean()
    q_norm_after = torch.norm(q_rot, dim=-1).mean()
    print(f"Query norm before rotation: {q_norm_before:.4f}")
    print(f"Query norm after rotation: {q_norm_after:.4f}")
    print("Note: RoPE preserves vector magnitude (rotation property)")

    # 5. Comparison
    print("\n5. Encoding Comparison")
    print("-" * 40)
    print("| Encoding   | Learnable | Relative | Extrapolation |")
    print("|------------|-----------|----------|---------------|")
    print("| Sinusoidal | No        | No       | Good          |")
    print("| Learned    | Yes       | No       | Poor          |")
    print("| Relative   | No        | Yes      | Good          |")
    print("| RoPE       | No        | Yes      | Excellent     |")

    print("\nRecommendations:")
    print("- Short fixed sequences: Sinusoidal or Learned")
    print("- Variable length: Relative")
    print("- Long sequences: RoPE")
    print("- Trading/Time series: RoPE + Calendar encoding")


if __name__ == "__main__":
    main()

File: 48_positional_encoding_timeseries/rust/examples/basic_encoding.rs
//! Example: Basic Positional Encoding
//!
//! This example demonstrates the fundamental positional encoding types:
//! sinusoidal, learned, relative, and rotary (RoPE).

use positional_encoding::{
    SinusoidalEncoding, LearnedEncoding, RelativeEncoding, RotaryEncoding,
    PositionalEncoding,
};
use ndarray::Array1;

fn main() {
    println!("Basic Positional Encoding Example");
    println!("==================================\n");

    // Sinusoidal Encoding
    println!("1. Sinusoidal Encoding");
    println!("-----------------------");

    let sinusoidal = SinusoidalEncoding::new(8, 100);
    let positions = vec![0, 1, 2, 5, 10];
    let encoded = sinusoidal.encode(&positions);

    println!("Dimension: {}", sinusoidal.dim());
    println!("Positions: {:?}", positions);
    for (i, pos) in positions.iter().enumerate() {
        let row: Vec<f64> = encoded.row(i).iter().map(|x| (x * 1000.0).round() / 1000.0).collect();
        println!("  pos {}: {:?}", pos, row);
    }

    // Key insight: Position 0 has sin(0)=0, cos(0)=1 pattern
    println!("\nNote: At position 0, even indices are 0 (sin), odd indices are 1 (cos)");

    // Learned Encoding
    println!("\n2. Learned Encoding");
    println!("-------------------");

    let learned = LearnedEncoding::new(8, 100);
    let learned_encoded = learned.encode(&[0, 1, 2]);

    println!("Learned embeddings (randomly initialized):");
    for i in 0..3 {
        let row: Vec<f64> = learned_encoded.row(i).iter().map(|x| (x * 100.0).round() / 100.0).collect();
        println!("  pos {}: {:?}", i, row);
    }
    println!("\nNote: These would be trained during model training");

    // Relative Encoding
    println!("\n3. Relative Encoding");
    println!("--------------------");

    let relative = RelativeEncoding::new(8, 50);

    // Show relative encodings between different positions
    let pairs = vec![(0, 5), (5, 0), (10, 15), (15, 10)];
    for (from, to) in pairs {
        let enc = relative.encode_relative(from, to);
        let row: Vec<f64> = enc.iter().map(|x| (x * 1000.0).round() / 1000.0).collect();
        println!("  {} -> {} (dist {}): {:?}", from, to, (to as i32 - from as i32).abs(), row);
    }
    println!("\nNote: Same distance = same encoding magnitude, direction differs by sign");

    // Rotary Encoding (RoPE)
    println!("\n4. Rotary Encoding (RoPE)");
    println!("-------------------------");

    let rope = RotaryEncoding::new(8, 100);

    // Apply rotation to a sample vector
    let x = Array1::from_vec(vec![1.0, 0.0, 0.5, 0.5, 0.0, 1.0, 0.5, 0.5]);
    println!("Original vector: {:?}", x.as_slice().unwrap());

    for pos in [0, 1, 5, 10] {
        let rotated = rope.apply_rotation(&x, pos);
        let row: Vec<f64> = rotated.iter().map(|v| (v * 1000.0).round() / 1000.0).collect();
        println!("  Rotated (pos {}): {:?}", pos, row);
    }

    // Verify norm preservation
    let orig_norm: f64 = x.iter().map(|v| v * v).sum::<f64>().sqrt();
    let rot_norm: f64 = rope.apply_rotation(&x, 50).iter().map(|v| v * v).sum::<f64>().sqrt();
    println!("\nNorm preservation: original={:.4}, rotated={:.4}", orig_norm, rot_norm);
    println!("Note: RoPE preserves vector magnitude (rotation property)");

    // Comparison
    println!("\n5. Encoding Comparison");
    println!("----------------------");
    println!("| Encoding   | Learnable | Relative | Extrapolation |");
    println!("|------------|-----------|----------|---------------|");
    println!("| Sinusoidal | No        | No       | Good          |");
    println!("| Learned    | Yes       | No       | Poor          |");
    println!("| Relative   | No        | Yes      | Good          |");
    println!("| RoPE       | No        | Yes      | Excellent     |");

    println!("\nRecommendations:");
    println!("- Short fixed sequences: Sinusoidal or Learned");
    println!("- Variable length: Relative");
    println!("- Long sequences: RoPE");
    println!("- Trading/Time series: RoPE + Calendar encoding");
}
